src/hierarchies/mfa_hierarchy.cc

```diff
@@ -0,0 +1,248 @@
+    }
+
+    else {
+        throw std::invalid_argument("Unrecognized hierarchy prior");
+    }
+}
+
+void MFAHierarchy::update_hypers(
+    const std::vector<bayesmix::AlgorithmState::ClusterState>& states) {
+    auto& rng = bayesmix::Rng::Instance().get();
+    if (prior->has_fixed_values()) {
+        return;
+    }
+
+    else {
+        throw std::invalid_argument("Unrecognized hierarchy prior");
+    }
+}
+
+void MFAHierarchy::update_summary_statistics(const Eigen::RowVectorXd& datum,
+    bool add) {
+    if (add) {
+        data_sum += datum;
+        data_sum_squares += datum .cwiseProduct(datum);
+        data.push_back(datum);
+    }
+    else {
+        data_sum -= datum;
+        data_sum_squares -= datum.cwiseProduct(datum);
+        auto pos = std::find(data.begin(), data.end(), datum)
+        data.erase(pos);
+    }
+}
+
+void MFAHierarchy::clear_summary_statistics() {
+    data_sum = Eigen::VectorXd::Zero(p);
+    data_sum_squares = Eigen::VectorXd::Zero(p);
+    data.clear();
+}
+
+void MFAHierarchy::set_state_from_proto(
+    const google::protobuf::Message& state_) {
+    auto& statecast = downcast_state(state_);
+    state.mu = statecast.mfa_state().mu(); // TODO add, ls??
+    state.psi = statecast.mfa_state().psi();
+    state.eta = statecast.mfa_state().eta();
+    state.lambda = statecast.mfa_state().lambda();
+    set_card(statecast.cardinality());
+}
+
+std::shared_ptr<bayesmix::AlgorithmState::ClusterState>
+MFAHierarchy::get_state_proto() const {
+    bayesmix::MFAState state_;
+    state_.set_mu(state.mu);
+    state_.set_psi(state.psi);
+    state_.set_Eta(state.Eta);
+    state_.set_Lambda(state.Lambda);
+
+    auto out = std::make_shared<bayesmix::AlgorithmState::ClusterState>();
+    out->mutable_mfa_state()->CopyFrom(state_);
+    return out;
+}
+
+void MFAHierarchy::set_hypers_from_proto(
+    const google::protobuf::Message& hypers_) {
+    auto& hyperscast = downcast_hypers(hypers_).mfa_state();
+    hypers->mutilde = hyperscast.mutilde();
+    hypers->alpha0 = hyperscast.alpha0();
+    hypers->beta = hyperscast.beta();
+    hypers->phi = hyperscast.phi();
+}
+
+std::shared_ptr<bayesmix::AlgorithmState::HierarchyHypers>
+MFAHierarchy::get_hypers_proto() const {
+    bayesmix::MFAState hypers_;
+    hypers_.set_mutilde(hypers->mutilde);
+    hypers_.set_alpha0(hypers->alpha0);
+    hypers_.set_beta(hypers->beta);
+    hypers_.set_phi(hypers->phi);
+
+    auto out = std::make_shared<bayesmix::AlgorithmState::HierarchyHypers>();
+    out->mutable_mfa_state()->CopyFrom(hypers_);
+    return out;
+}
+
+void MFAHierarchy::sample_full_cond(bool update_params = true) override {
+    if (this->card == 0) {
+        // No posterior update possible
+        static_cast<Derived*>(this)->sample_prior();
+    }
+    else {
+        std::shared_ptr<State> newstate = nullptr;
+
+        if (update_params)
+            newstate = &state;
+        else
+            newstate = new State;
+
+        sample_Eta(*newstate);
```

Comment:
rewrite these methods so that they don't take a reference to a state, but update the current one